### ISSUE TITLE
NotificationConfiguration is not a valid property of AWS::AutoScaling::AutoScalingGroup

### ISSUE THREAD

**Issue by aarongorka** [martysweet/cfn-lint#98]
```
  AutoscalingGroup:
    Type: AWS::AutoScaling::AutoScalingGroup
    Properties:
      AvailabilityZones:
        - ap-southeast-2a
      NotificationConfiguration:
        TopicARN: !Ref 'SnsAutoscaling'
        NotificationTypes:
          - autoscaling:EC2_INSTANCE_LAUNCH
          - autoscaling:EC2_INSTANCE_LAUNCH_ERROR
          - autoscaling:EC2_INSTANCE_TERMINATE
          - autoscaling:EC2_INSTANCE_TERMINATE_ERROR
      HealthCheckType: ELB
      HealthCheckGracePeriod: '900'
...
```

This valid template returns an error when validating. The AWS documentation seems to be inconsistent as to what the property is actually called:

NotificationConfiguration**s**
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html

NotificationConfiguration (note the random s in the JSON string - something went wrong here!)
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-notificationconfigurations.html

As a result, data/aws_resources_specification.json has both singular and plural forms:

```json
    "AWS::AutoScaling::AutoScalingGroup.NotificationConfiguration": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-notificationconfigurations.html",
      "Properties": {
        "NotificationTypes": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-notificationconfigurations.html#cfn-as-group-notificationconfigurations-notificationtypes",
          "DuplicatesAllowed": true,
          "PrimitiveItemType": "String",
          "Required": false,
          "Type": "List",
          "UpdateType": "Mutable"
        },
        "TopicARN": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-notificationconfigurations.html#cfn-autoscaling-autoscalinggroup-notificationconfigurations-topicarn",
          "PrimitiveType": "String",
          "Required": true,
          "UpdateType": "Mutable"
        }
      }
```

**Comment by akdor1154** [comment]
I think `cfn-lint` is in the clear here.
`NotificationConfigurations` is the name of the property AWS wants to see on an `AutoScalingGroup`.
`NotificationConfiguration` is the name of the schema / data-type that the value of that property should take.

According to the AWS docs, your template should look like
```yaml
  AutoscalingGroup:
    Type: AWS::AutoScaling::AutoScalingGroup
    Properties:
      AvailabilityZones:
        - ap-southeast-2a
      NotificationConfigurations:
        - TopicARN: !Ref 'SnsAutoscaling'
          NotificationTypes:
            - autoscaling:EC2_INSTANCE_LAUNCH
            - autoscaling:EC2_INSTANCE_LAUNCH_ERROR
            - autoscaling:EC2_INSTANCE_TERMINATE
            - autoscaling:EC2_INSTANCE_TERMINATE_ERROR
      HealthCheckType: ELB
      HealthCheckGracePeriod: '900'
```
This template passes `cfn-lint`. (I haven't deployed it though, let me check this)

**Comment by akdor1154** [comment]
The above template (expanded slightly: 
```yaml
Resources:

  Topic: 
    Type: AWS::SNS::Topic

  AutoscalingGroup:
    Type: AWS::AutoScaling::AutoScalingGroup
    Properties:
      MaxSize: 0
      MinSize: 0
      LaunchConfigurationName: !Ref LC
      AvailabilityZones:
        - ap-southeast-2a
      TagProperty:
        Key: test
        Value: val
      NotificationConfigurations:
        - TopicARN: !Ref Topic
          NotificationTypes:
            - autoscaling:EC2_INSTANCE_LAUNCH
            - autoscaling:EC2_INSTANCE_LAUNCH_ERROR
            - autoscaling:EC2_INSTANCE_TERMINATE
            - autoscaling:EC2_INSTANCE_TERMINATE_ERROR
      HealthCheckType: ELB
      HealthCheckGracePeriod: '900'

  LC:
    Type: AWS::AutoScaling::LaunchConfiguration
    Properties:
      ImageId: ami-ff4ea59d
      InstanceType: t2.micro
```
) launches successfully.

I would consider the fact that `NotificationConfiguration` works in practice to be an undocumented feature, maybe even implementation detail, in CloudFormation.

**Comment by aarongorka** [comment]
I'm happy with that, I will fix my templates 🙂
